File: ktox-lua-cc/src/main/kotlin/programs/GhFetch.kt
package programs

import common.fsMakeDir
import common.ktoxDownloadFile
import common.ktoxDownloadFileText

// Pulls Digsite and its dependencies onto a fresh turtle/computer, straight
// from GitHub. Each entry is a path relative to the repo base and also the
// local path it's saved to, preserving directory structure — unlike
// moonman's `sync`, which flattens every file to the CC root via
// fs.getName() and would break require("lib/Movement") the moment this
// package's files land there. See AGENTS.md.
//
// The actual file list lives in FILES_MANIFEST (fetched, not hardcoded
// here) — see that constant's own comment for why, and MANIFEST.md at
// the repo root for how to maintain it. This file only knows the ONE
// thing that can't itself come from the manifest: where to find it.
//
// DEFAULT_BRANCH tracks whichever branch is under active iteration —
// currently feat/ktox-lua-storage. main doesn't have
// src/pkg/player/8durt yet, so this can't point there — move it to
// "main" (matching moonman's manifest convention) once that merges.
//
// Usage: ghfetch [branch]
//   - branch (optional): git ref to fetch from, e.g. a feature branch
//     you're iterating on. Defaults to DEFAULT_BRANCH when omitted.

const val DEFAULT_BRANCH = "feat/ktox-lua-storage"
const val FILES_MANIFEST = "files.manifest"

fun main(args: Array<String>) {
    val branch = if (args.size >= 1) args[1] else DEFAULT_BRANCH
    val repoBase = "https://raw.githubusercontent.com/caboose1029/minecraft-cc/${branch}/src/pkg/player/8durt"
    println("Fetching from branch: ${branch}")

    val manifestUrl = "${repoBase}/${FILES_MANIFEST}"
    val manifestText = ktoxDownloadFileText(manifestUrl)
    if (manifestText == "MISSING") {
        println("Could not fetch the file manifest (${FILES_MANIFEST}) - aborting.")
        return
    }
    // Also save it to disk, same as every other synced file - not
    // strictly needed for this run (already have the text above), but
    // keeps it inspectable and consistent with everything else ghfetch
    // deploys.
    ktoxDownloadFile(manifestUrl, FILES_MANIFEST)

    // One path per line. A trailing blank line (a file conventionally
    // ends with one) is skipped below, not an error.
    val files = manifestText.split("\n")

    var i = 1
    var failures = 0
    var fetched = 0
    while (i <= files.size) {
        val path = files[i]
        if (path != "") {
            // Every real path in this tree is either bare ("Foo.lua") or
            // exactly one directory deep ("lib/Foo.lua") - never nested
            // further, so the directory is just the first "/"-separated
            // part when there is one.
            val pathParts = path.split("/")
            if (pathParts.size >= 2) {
                fsMakeDir(pathParts[1])
            }
            val url = "${repoBase}/${path}"
            println("Fetching ${path}...")
            val ok = ktoxDownloadFile(url, path)
            if (ok) {
                println("  ok")
                fetched += 1
            } else {
                println("  FAILED: ${path}")
                failures += 1
            }
        }
        i += 1
    }

    if (failures > 0) {
        println("${failures} file(s) failed.")
    } else {
        println("Fetched ${fetched} file(s).")
    }
}

File: src/pkg/player/8durt/GhFetch.lua
-- package: programs

require("ktox-lib")
ktox_sourcemap_traceback(debug and debug.getinfo and (debug.getinfo(1) or {}).short_src or "", "GhFetch.kt", {["1-11"]=1,["12"]=32,["13"]=33,["14"]=34,["15"]=36,["16"]=37,["17"]=38,["18"]=39,["19-20"]=40,["21"]=46,["22"]=50,["23"]=52,["24"]=53,["25"]=54,["26"]=55,["27"]=56,["28"]=57,["29"]=62,["30"]=63,["31-32"]=64,["33"]=66,["34"]=67,["35"]=68,["36"]=69,["37"]=70,["38-39"]=71,["40"]=73,["41-43"]=74,["44-45"]=77,["46"]=80,["47-48"]=81,["49-52"]=83}, "programs")

DEFAULT_BRANCH = "feat/ktox-lua-storage"

FILES_MANIFEST = "files.manifest"

---@param args table
function main(args)
    local branch = (#(args) >= 1 and args[1] or DEFAULT_BRANCH)
    local repoBase = "https://raw.githubusercontent.com/caboose1029/minecraft-cc/" .. tostring(branch) .. "/src/pkg/player/8durt"
    println("Fetching from branch: " .. tostring(branch))
    local manifestUrl = tostring(repoBase) .. "/" .. tostring(FILES_MANIFEST)
    local manifestText = ktoxDownloadFileText(manifestUrl)
    if manifestText == "MISSING" then
        println("Could not fetch the file manifest (" .. tostring(FILES_MANIFEST) .. ") - aborting.")
        return
    end
    ktoxDownloadFile(manifestUrl, FILES_MANIFEST)
    local files = ktox_split(manifestText, "\n")
    local i = 1
    local failures = 0
    local fetched = 0
    while i <= #(files) do
        local path = files[i]
        if path ~= "" then
            local pathParts = ktox_split(path, "/")
            if #(pathParts) >= 2 then
                fs.makeDir(pathParts[1])
            end
            local url = tostring(repoBase) .. "/" .. tostring(path)
            println("Fetching " .. tostring(path) .. "...")
            local ok = ktoxDownloadFile(url, path)
            if ok then
                println("  ok")
                fetched = ktox_plusAssign(fetched, 1)
            else
                println("  FAILED: " .. tostring(path))
                failures = ktox_plusAssign(failures, 1)
            end
        end
        i = ktox_plusAssign(i, 1)
    end
    if failures > 0 then
        println(tostring(failures) .. " file(s) failed.")
    else
        println("Fetched " .. tostring(fetched) .. " file(s).")
    end
end


main({...})
